lib/Dialect/TritonGPU/Transforms/FP32DotTC.cpp

```diff
@@ -0,0 +1,93 @@
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+
+using namespace mlir;
+namespace tt = mlir::triton;
+
+namespace {
+
+// Implement 3xTF32 trick https://github.com/NVIDIA/cutlass/discussions/385
+// For a, b fp32
+// dot(a, b, f32Backend="3xtf32") ->
+//  let aBig = f32ToTF32(a), aSmall = a - aBig;
+//  let bBig = f32ToTF32(b), bSmall = b - bBig;
+//  dot(aSmall, bBig, f32Backend="tf32") +
+//  dot(aBig, bSmall, f32Backend="tf32") +
+//  dot(aBig, bBig, f32Backend="tf32")
+class TF32x3 : public OpRewritePattern<tt::DotOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult
+  matchAndRewrite(tt::DotOp dotOp,
+                  mlir::PatternRewriter &rewriter) const override {
+
+    auto isF32 = [](mlir::Value operand) {
+      return operand.getType()
+          .cast<RankedTensorType>()
+          .getElementType()
+          .isF32();
+    };
+
+    if (!(dotOp.getF32Backend() == "3xtf32" && isF32(dotOp.getOperand(0)) &&
```

Comment:
You can do dotOp.getA/B/C().  These are also always RankedTensorType'ed values, so you can leave off a cast in some places.